flink-libraries/flink-cep/src/test/java/org/apache/flink/cep/nfa/NFAIterativeConditionTimerContextTest.java

```diff
@@ -0,0 +1,152 @@
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.apache.flink.cep.nfa;
+
+import org.apache.flink.cep.Event;
+import org.apache.flink.cep.pattern.Pattern;
+import org.apache.flink.cep.pattern.conditions.IterativeCondition;
+import org.apache.flink.cep.utils.NFATestHarness;
+import org.apache.flink.cep.utils.TestCepTimerService;
+import org.apache.flink.streaming.runtime.streamrecord.StreamRecord;
+import org.apache.flink.util.TestLogger;
+
+import org.junit.Assert;
+import org.junit.Test;
+
+import java.util.Collections;
+import java.util.List;
+
+import static org.apache.flink.cep.utils.EventBuilder.event;
+import static org.apache.flink.cep.utils.NFATestHarness.forPattern;
+import static org.apache.flink.cep.utils.NFATestUtilities.compareMaps;
+import static org.hamcrest.Matchers.is;
+import static org.hamcrest.Matchers.nullValue;
+
+/**
+ * Tests for accesing time properties from {@link IterativeCondition}.
+ */
+public class NFAIterativeConditionTimerContextTest extends TestLogger {
+
+	@Test
+	public void testEventTimestampInEventTime() throws Exception {
+		Event event = event().withId(1).build();
+
+		Pattern<Event, ?> pattern = Pattern.<Event>begin("start").where(new IterativeCondition<Event>() {
+			@Override
+			public boolean filter(Event value, Context<Event> ctx) throws Exception {
+				return ctx.timestamp() != null && ctx.timestamp() == 3;
+			}
+		});
+
+		TestCepTimerService cepTimerService = new TestCepTimerService();
+		cepTimerService.setProcessingTime(false);
+		NFATestHarness nfaTestHarness = forPattern(pattern)
+			.withTimerService(cepTimerService)
+			.build();
+
+		List<List<Event>> resultingPatterns = nfaTestHarness.feedRecord(new StreamRecord<>(event, 3));
+
+		compareMaps(resultingPatterns, Collections.singletonList(
+			Collections.singletonList(event)
+		));
+	}
+
+	@Test
+	public void testEventTimestampInProcessingTime() throws Exception {
+		Event event = event().withId(1).build();
+
+		Pattern<Event, ?> pattern = Pattern.<Event>begin("start").where(new IterativeCondition<Event>() {
+			@Override
+			public boolean filter(Event value, Context<Event> ctx) throws Exception {
+				Assert.assertThat(ctx.timestamp(), is(nullValue()));
+				return true;
+			}
+		});
+
+		TestCepTimerService cepTimerService = new TestCepTimerService();
+		cepTimerService.setProcessingTime(true);
+		NFATestHarness nfaTestHarness = forPattern(pattern)
+			.withTimerService(cepTimerService)
+			.build();
+
+		List<List<Event>> resultingPatterns = nfaTestHarness.feedRecord(new StreamRecord<>(event, 3));
+
+		compareMaps(resultingPatterns, Collections.singletonList(
+			Collections.singletonList(event)
+		));
+	}
+
```

Comment:
Same as above, the following two tests share a big part of the code which we could extract in a separate method.